### ISSUE TITLE
Allow conditional exports (or sub-exports)

### ISSUE THREAD

**Issue by Tyrannosaurus1234** [godotengine/godot-proposals#2582]
### Describe the project you are working on
Text-based RPG

### Describe the problem or limitation you are having in your project
I have a "door" class that exports a number of variables to the editor for easy configuration, relevant to this topic I have:
`pickable: bool`
`pickLevel: int`

pickLevel is useless if the door is not pickable, and only serves to clutter the script exports dock thing in the editor. For one variable this isn't a big deal, but if I were to add other things like kicking down doors or casting spells to unlock them, each depending on a different stat, it would quickly get out of hand.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Allow exporting variables conditionally, or attaching exports to other exports, such that `pickLevel` is only exposed to the editor if `pickable` is True.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
I don't know how to cleanly express this in code, but maybe something like this:
`export(bool) var pickable
export(int, pickable) var pickLevel`
or maybe something with decorators?
`export(bool) var pickable
@export_pickable
export(int) var pickLevel`

### If this enhancement will not be used often, can it be worked around with a few lines of script?
As far as I'm aware there is no way to achieve this functionality currently.

### Is there a reason why this should be core and not an add-on in the asset library?
It changes the core export system.

**Comment by KoBeWi** [comment]
This can be achieved with `_get_property_list()`.

**Comment by Xrayez** [comment]
Here's an example project showcasing above:

**Test project:**
[conditional-export-property-list.zip](https://github.com/godotengine/godot-proposals/files/6290652/conditional-export-property-list.zip)

```gdscript
tool
extends Node2D

export(bool) var pickable = false setget set_pickable
var pick_level: int

func set_pickable(p_pickable):
	pickable = p_pickable
	property_list_changed_notify()

func _get_property_list() -> Array:
	var list = []
	if pickable:
		list.push_back(
			{ name = "pick_level", type = TYPE_INT }
		)
	return list

```

**Comment by MarcusElg** [comment]
Maybe there could be a @export_conditionally(expression) annotation? So your example would be:
`
@export var pickable: Bool
@export_conditionally(pickable == true) var pickLevel: Int
`

I feel it would be useful to hide properties that depend on other, to avoid confusing users with properties that don't affect anything.

**Comment by Calinou** [comment]
> Maybe there could be a @export_conditionally(expression) annotation? So your example would be:
> 
> ```
> @export var pickable: Bool
> @export_conditionally(pickable == true) var pickLevel: Int
> ```
> 
> I feel it would be useful to hide properties that depend on other properties, to avoid confusing users with properties that don't affect anything.

One issue with `@export_conditionally` is that it *implies* the property is not exported if the condition isn't met. In this case, this means its value is lost in scene files when saving the scene if the condition isn't met at that time.

We should probably ensure hidden properties are always saved by default to avoid data loss.

**Comment by MarcusElg** [comment]
> > Maybe there could be a @export_conditionally(expression) annotation? So your example would be:
> > ```
> > @export var pickable: Bool
> > @export_conditionally(pickable == true) var pickLevel: Int
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I feel it would be useful to hide properties that depend on other properties, to avoid confusing users with properties that don't affect anything.
> 
> One issue with `@export_conditionally` is that it _implies_ the property is not exported if the condition isn't met. In this case, this means its value is lost in scene files when saving the scene if the condition isn't met at that time.
> 
> We should probably ensure exported-but-hidden properties are always saved by default to avoid data loss.

Yes, data loss should be avoided when possible. Maybe some name with visibility instead? As it's not visible in the inspector, something like ```@conditionally_visible```? Or maybe that's too inconsistent with existing naming.

Regardless of the name I feel like it would be a very useful feature to have as it's quite a common usecase for custom inspectors, and it would be nice to avoid having to override the property list for it.

**Comment by KoBeWi** [comment]
If the property is conditional and the conditional is not met, it means you don't need it and there is no data loss. Property value is chached internally until scene is reloaded, so accidental data loss is unlikely too.

**Comment by MarcusElg** [comment]
I wonder if this would be useful for the editor itself. Right now the Button's icon alignment property is shown for example regardless of a icon is selected or not. Showing properties that don't affect anything is defiently confusing for the user.

**Comment by KoBeWi** [comment]
We already use conditional properties in the editor. This proposal is about GDScript convenience, which wouldn't have any use in the editor.

**Comment by YuriSizov** [comment]
This is actually a duplicate of https://github.com/godotengine/godot-proposals/issues/1056. The OP there proposes a different solution, but an annotation is more likely to be accepted if we implement this (and also proposed in the comments).
